eng/release/DiagnosticsReleaseTool/Common/AzureBlobPublisher.cs

```diff
@@ -0,0 +1,267 @@
+                        Identifier = AccessPolicyDownloadId,
+                        Protocol = SasProtocol.Https,
+                    };
+                    Uri accessUri = blobClient.GenerateSasUri(sasBuilder);
+
+                    using BlobDownloadStreamingResult blobStream = (await blobClient.DownloadStreamingAsync(cancellationToken: ct)).Value;
+                    srcStream.Position = 0;
+                    completed = await VerifyFileStreamsMatchAsync(srcStream, blobStream, ct);
+
+                    result = accessUri;
+                }
+                catch (IOException ioEx) when (!(ioEx is PathTooLongException))
+                {
+                    _logger.LogWarning(ioEx, $"Failed to publish {fileMap.LocalSourcePath}, retries remaining: {retriesLeft}.");
+
+                    /* Retry IO exceptions */
+                    retriesLeft--;
+                    loopDelay *= 2;
+
+                    if (retriesLeft > 0)
+                    {
+                        await Task.Delay(loopDelay, ct);
+                    }
+                }
+                catch (Exception ex)
+                {
+                    // Azure errors have their own built-in retry logic, so just abort if we got an AzureResponseException
+                    _logger.LogWarning(ex, $"Failed to publish {fileMap.LocalSourcePath}, unexpected error, aborting.");
+                    return null;
+                }
+            } while (retriesLeft > 0 && !completed);
+
+            return result?.OriginalString;
+        }
+
+        private static string GetBlobName(string releaseName, string relativeFilePath)
+        {
+            return $"{releaseName}/{relativeFilePath}";
+        }
+
+        private async Task<BlobContainerClient> GetClient(CancellationToken ct)
+        {
+            if (_client == null)
+            {
+                BlobServiceClient serviceClient = new BlobServiceClient(AccountBlobUri, AccountCredential, BlobOptions);
+                _logger.LogInformation($"Attempting to connect to {serviceClient.Uri} to store blobs.");
+
+                BlobContainerClient newClient;
+                int attemptCt = 0;
+                do
+                {
+                    try
+                    {
+                        newClient = (await serviceClient.CreateBlobContainerAsync(_containerName, PublicAccessType.None, metadata: null, ct)).Value;
+                        _logger.LogInformation($"Created new contianer: {_containerName}.");
+                    }
+                    // 409 is the error code for an existing blob container already existing
+                    catch (RequestFailedException rfe) when (rfe.Status == 409)
+                    {
+                        _logger.LogInformation($"Found existing contianer: {_containerName}.");
+                        // Got an existing container
+                        newClient = serviceClient.GetBlobContainerClient(_containerName);
```

Comment:
nit: Should this be done first before trying to create the container? I think there's an ExistsAsync method you can use to check that it truly exists before falling back to CreateBlobClientAsync

Reply:
The ExistsAsync doesn't exist in this API (I'm using the blob client not the storage account client).

Reply:
```cs
newClient = serviceClient.GetBlobContainerClient(_containerName);
if (!await newClient.ExistsAsync(...))
{
    newClient = await (serviceClient.CreateBlobContainerAsync(_containerName, PublicAccessType.None, metadata: null, ct)).Value;
}
```
https://docs.microsoft.com/en-us/dotnet/api/azure.storage.blobs.blobserviceclient.getblobcontainerclient?view=azure-dotnet
https://docs.microsoft.com/en-us/dotnet/api/azure.storage.blobs.blobcontainerclient.existsasync?view=azure-dotnet#Azure_Storage_Blobs_BlobContainerClient_ExistsAsync_System_Threading_CancellationToken_
https://docs.microsoft.com/en-us/dotnet/api/azure.storage.blobs.blobserviceclient.createblobcontainerasync?view=azure-dotnet

Reply:
Ahh, I didn't realize I the client creation didn't result in an API call to azure.